### PR TITLE
fix: DwC Auth Token not available

### PR DESCRIPTION
**Pull request by MatKuhr**
## Context

SAP/cloud-sdk-java-backlog#424.

I think we just forgot to implement this earlier. This change makes our `AuthTokenAccessor` actually return a token in the DwC case. Without this change, one always gets an exception (unless using CAP).

## Definition of Done

- [x] Functionality scope stated & covered
- [x] Tests cover the scope above
- [x] ~Error handling created / updated & covered by the tests above~
- [x] ~Documentation updated~
- [x] Release notes updated

### CODE REVIEW

File: cloudplatform/security/src/main/resources/META-INF/services/com.sap.cloud.sdk.cloudplatform.security.secret.SecretStoreFacade [hunk 1]
```diff
@@ -1 +0,0 @@
-com.sap.cloud.sdk.cloudplatform.security.secret.ScpCfSecretStoreFacade
```
Comment (MatKuhr) [inline cloudplatform/security/src/main/resources/META-INF/services/com.sap.cloud.sdk.cloudplatform.security.secret.SecretStoreFacade:1]: driveby cleanup of dead code

File: cloudplatform/security/src/main/resources/META-INF/services/com.sap.cloud.sdk.cloudplatform.security.AuthTokenFacade [hunk 2]
```diff
@@ -1 +0,0 @@
-com.sap.cloud.sdk.cloudplatform.security.DefaultAuthTokenFacade
```
Comment (MatKuhr) [inline cloudplatform/security/src/main/resources/META-INF/services/com.sap.cloud.sdk.cloudplatform.security.AuthTokenFacade:1]: removing this should be fine because the Accessor uses this facade regardless by default (without classloading)

Comment (CharlesDuboisSAP) [reply cloudplatform/security/src/main/resources/META-INF/services/com.sap.cloud.sdk.cloudplatform.security.AuthTokenFacade:1]: It doesn't if the `DwcAuthTokenFacade` is class loaded

Comment (CharlesDuboisSAP) [reply cloudplatform/security/src/main/resources/META-INF/services/com.sap.cloud.sdk.cloudplatform.security.AuthTokenFacade:1]: It doesn't if the `DwcAuthTokenFacade` is class loaded

**Review by CharlesDuboisSAP — COMMENTED** [review]
Should this be E2E tested?

File: cloudplatform/connectivity-dwc/pom.xml [hunk 3]
```diff
@@ -69,6 +69,10 @@
 			<groupId>org.slf4j</groupId>
 			<artifactId>slf4j-api</artifactId>
 		</dependency>
+		<dependency>
+			<groupId>com.auth0</groupId>
+			<artifactId>java-jwt</artifactId>
+		</dependency>
```
Comment (newtork) [inline cloudplatform/connectivity-dwc/pom.xml:75]: (Question)

This is not "test" scoped, because it comes in "compile" transitively, right?

Comment (MatKuhr) [reply cloudplatform/connectivity-dwc/pom.xml:75]: not transitive, otherwise it wouldn't be (allowed to be) listed here. Usage is [here](https://github.com/SAP/cloud-sdk-java/pull/337/files#diff-5295696d29a6734eb094fcad0b32195dadfc42918bcd7c3e153e2a05f219b653R6).

File: cloudplatform/connectivity-dwc/src/main/java/com/sap/cloud/sdk/cloudplatform/security/DwcAuthTokenFacade.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    @Nonnull
+    @Override
+    public Try<AuthToken> tryGetCurrentToken()
+    {
+        @Nullable
+        final ThreadContext currentContext = ThreadContextAccessor.getCurrentContextOrNull();
+        if( currentContext != null && currentContext.containsProperty(AUTH_TOKEN) ) {
+            return currentContext.getPropertyValue(AUTH_TOKEN);
+        }
+        return Try.of(DwcAuthTokenFacade::extractAuthTokenFromDwcHeaders);
```
Comment (newtork) [inline cloudplatform/connectivity-dwc/src/main/java/com/sap/cloud/sdk/cloudplatform/security/DwcAuthTokenFacade.java:30]: (Minor)

Please note: not storing the potential exception in thread context will lead to multiple/many exception instantiations (2x per `tryGetCurrentToken`).

**Review by MatKuhr — COMMENTED** [review]
currently drafting an E2E test [here](https://github.tools.sap/cloudsdk/cloud-sdk-java-tests/pull/44) based on [this test code](https://github.tools.sap/cloudsdk/dwc-cap-orbits/commit/ac9724b6c96a261bcee6fa8c11eb014254990a1b).

File: cloudplatform/connectivity-dwc/src/main/java/com/sap/cloud/sdk/cloudplatform/security/DwcAuthTokenFacade.java (hunk 4 above)
Comment (newtork) [reply cloudplatform/connectivity-dwc/src/main/java/com/sap/cloud/sdk/cloudplatform/security/DwcAuthTokenFacade.java:30]: I'm not sure we need to address this right here/now.

File: cloudplatform/connectivity-dwc/src/main/java/com/sap/cloud/sdk/cloudplatform/DwcHeaderUtils.java [hunk 5]
```diff
@@ -85,6 +89,20 @@ public static String getDwcPrincipalIdOrThrow()
...
     }
 
+    /**
+     * This method fetches the value of the {@link #DWC_JWT_HEADER} header or throws an
+     * {@link DwcHeaderNotFoundException} if the header was not found.
+     *
+     * @return The value of the {@link #DWC_JWT_HEADER} header.
+     * @throws DwcHeaderNotFoundException
+     *             if the header was not found.
+     */
```
Comment (Johannes-Schneider) [inline cloudplatform/connectivity-dwc/src/main/java/com/sap/cloud/sdk/cloudplatform/DwcHeaderUtils.java:99]: [minor]

```suggestion
     * @since 5.6.0
     */
```

File: cloudplatform/connectivity-dwc/src/main/java/com/sap/cloud/sdk/cloudplatform/DwcHeaderUtils.java [hunk 6]
```diff
@@ -85,6 +89,20 @@ public static String getDwcPrincipalIdOrThrow()
...
+    @Nonnull
+    public static String getDwcTokenOrThrow()
```
Comment (Johannes-Schneider) [inline cloudplatform/connectivity-dwc/src/main/java/com/sap/cloud/sdk/cloudplatform/DwcHeaderUtils.java:101]: [minor]

For consistency...?

```suggestion
    public static String getDwcJwtOrThrow()
```

File: cloudplatform/connectivity-dwc/src/main/java/com/sap/cloud/sdk/cloudplatform/security/DwcAuthTokenFacade.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+import com.sap.cloud.sdk.cloudplatform.DwcHeaderUtils;
+import com.sap.cloud.sdk.cloudplatform.security.exception.AuthTokenAccessException;
+import com.sap.cloud.sdk.cloudplatform.thread.ThreadContext;
+import com.sap.cloud.sdk.cloudplatform.thread.ThreadContextAccessor;
+
+import io.vavr.control.Try;
+
+/**
+ * Auth token facade for Deploy with Confidence (DwC) environment.
+ */
```
Comment (Johannes-Schneider) [inline cloudplatform/connectivity-dwc/src/main/java/com/sap/cloud/sdk/cloudplatform/security/DwcAuthTokenFacade.java:16]: [minor]

```suggestion
  * @since 5.6.0
  */
```

File: cloudplatform/connectivity-dwc/src/test/java/com/sap/cloud/sdk/cloudplatform/security/DwcAuthTokenFacadeTest.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+import io.vavr.control.Try;
+
+class DwcAuthTokenFacadeTest
+{
+    @Test
+    void testFacadeIsPickedUpAutomatically()
+    {
+        assertThat(AuthTokenAccessor.getAuthTokenFacade()).isInstanceOf(DwcAuthTokenFacade.class);
+    }
```
Comment (Johannes-Schneider) [inline cloudplatform/connectivity-dwc/src/test/java/com/sap/cloud/sdk/cloudplatform/security/DwcAuthTokenFacadeTest.java:27]: 👍

File: cloudplatform/connectivity-dwc/src/main/java/com/sap/cloud/sdk/cloudplatform/security/DwcAuthTokenFacade.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+            return currentContext.getPropertyValue(AUTH_TOKEN);
+        }
+        return Try.of(DwcAuthTokenFacade::extractAuthTokenFromDwcHeaders);
+    }
+
+    @Nonnull
+    private static AuthToken extractAuthTokenFromDwcHeaders()
+    {
+        try {
+            final String token = DwcHeaderUtils.getDwcTokenOrThrow();
```
Comment (Johannes-Schneider) [inline cloudplatform/connectivity-dwc/src/main/java/com/sap/cloud/sdk/cloudplatform/security/DwcAuthTokenFacade.java:37]: [question]

Can you remind me what we decided regarding "DwC header injection" attacks (i.e. someone somehow calls the app and injects their own `dwc-*` headers)?
I know we discussed this topic a while back but cannot quite recall what we decided.

Comment (MatKuhr) [reply cloudplatform/connectivity-dwc/src/main/java/com/sap/cloud/sdk/cloudplatform/security/DwcAuthTokenFacade.java:37]: we figured not an issue for various reasons, including for example:
- the dwc dependencies don't come in automatically and it's decently unlikely that they can come transitively outside of DwC
- even if they are present, the multi tenancy would not work at all without the dwc headers, so any multi-tenant app should detect this during development
